File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res=new ArrayList<>();
        pragenerator(res,"",0,0,n);
        return res;
    }
    public static void  pragenerator(List<String> res,String str,int o,int c,int n){
        if(o+c==n*2){
            res.add(str);
            return;
        }
        if(o<n){
            pragenerator(res,str+"(",o+1,c,n);
        }
        if(c<o){
            pragenerator(res,str+")",o,c+1,n);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        IList<string> ls=new List<string>();
            back(n,0,0,ls,"");
            return ls;
        
    }
    public static void back(int n,int o,int c,IList<string> list,string str){
        if(str.Length==n*2){
            list.Add(str);
            return;
        }
        if(o<n){
            back(n,o+1,c,list,str+'(');
        }
        if(c<o){
            back(n,o,c+1,list,str+')');
        }
    }
}